## Results

### Scaling up an interpretable perturbation biology model to genome scales

A variety of computational methods have been developed to infer cellular networks from molecular profiling efforts. Most often these have been descriptive statistical models, such as co-expression networks or mutual information methods, which identify network connections but cannot make specific predictions for the effect of certain perturbations. 


Given these limitations, we applied a recently-proposed approach, in which a non-linear system of ordinary differential equations is fit to a panel of molecular intervention and profiling experiments. By doing so, the model is both able to predict the effect of new interventions alone or in combination, and reconstructs a network of interactions among individual molecular nodes.

While a variety of techniques exist to improve the scalability of ODE models, we found that the CellBox approach was still limiting at genome scales. Briefly, each iteration requires both initial solving and sensitivity calculations for each perturbation applied. One model solution includes a number of dimensions equal to the number of molecular nodes within the model, and the number of parameters scales with the square of the number of species. Because of the parameter scaling, careful selection of regularization will be necessary for larger models as they in general will be underdetermined without these constraints. Finally, the fitting process requires many iterations which increases with the number of unknown parameters, and local minima are sure to hinder gradient-based optimization.

As a result of these limitations, we explored alternative approaches to fitting such a model. We were interested in exploring the XXX, wherein 1000 genes were each knocked down by siRNA or CRISPR, and then the expression of all 1000 genes was profiled. As a result we were left with 1,000,000 training points for capturing the effect of gene-gene interactions. As in earlier work, we had only a single snapshot measurement, and so assumed that our measurements were made at steady-state. Thus, a first simplification we could make was to dispell with ODE solving and consider the deviation of our differential equations from steady-state as our fitting cost function. This also allowed us to recognize that all our perturbations could be expressed as a single matrix equation for "fused" solving.

Expressing the steady-state CellBox model as a single matrix equation then revealed an iterative solving routine we could apply for more efficient fitting. In other fitting models with multiple unknowns, such as non-negative matrix factorization or tensor factorization, one can alternate between solving for the best-fit value of each unknown. We applied this by calculating the best-fit $\epsilon$, then using the least-squares solution of the interation matrix $W$, and alternating between these steps until convergence. Importantly, this iterative scheme, we term FactorBox, still allows for various regularization such as orthogonality, sparsity, and other forms.

![**Figure 1.** Text.](figure1.svg "Figure 1"){#fig:quantFig width="100%"}

### FactorBox reduces cellular context-dependent effects

With a working scalable model of gene-gene interaction effects, we wished to determine whether FactorBox identified consistent results that were independent of cellular context. To do so, we fit the model to the same knockout and profiling experiments performed in a series of different cell lines. We then quantified the difference in the inferred gene-gene interaction network between each cell line. As a baseline of comparison, we used the gene loadings matrix from principal components analysis of the same experiment. We observed strong concordance in the inferred interaction effects, and greater agreement across cell lines in the FactorBox model as compared to the PCA loadings. This indicates that the baseline expression and therefore the consequent gene expression patterns are different between cell lines, but the individual interactions effects are shared.

With this in mind, we devised a modified scheme to fit the gene-gene interaction effects taking into account all of the cell lines together. By doing so, we enforced that all cell line responses should be explained by a common set of gene-gene interactions. Despite adding ~8,000,000 measurements and only ~8,000 new parameters, the model fits were of equal quality to those obtained with each cell line separately. This suggests that a common gene-gene interaction network can be obtained, and cell line context differences can be explained through non-linearies introducted through baseline expression differences. (For example, if a gene is not expressed, the influence of the pathway it is in will be terminated at that node.)

![**Figure 2.** Text.](figure2.svg "Figure 2"){#fig:fig2 width="100%"}

### FactorBox provides a de novo, quantitative, directed graph of gene-gene effects

With confidence that FactorBox provides a reliable network of gene-gene interactions effects, we sought some preliminary exploration of the inferred network. Gene interactions formed a scale-free network, with extremes of highly connected and isolated nodes. Importantly, the perturbation data makes it possible to identify loops of interactions, and we observed quite a few positive and negative feedback loops across the graph. As just one example...

![**Figure 3.** Text.](figure3.svg "Figure 3"){#fig:fig3 width="100%"}

### Variance flux analysis identifies the drivers of cell-to-cell variation

Compare the eigenvectors of the graph to the eigenvectors of the covariance matrix.


![**Figure 4.** Text.](figure4.svg "Figure 4"){#fig:fig4 width="100%"}
